### ISSUE TITLE
Better error on doc comments preceding implicit returns

### ISSUE THREAD

**Issue by awr1** [nim-lang/Nim#12093]
```nim
proc foo*(): bool =
  ## doc comment
  true

proc bar*(): bool =
  foo()
  ## doc comment
```
Fails with:
```
/usercode/in.nim(7, 6) Error: expression 'foo()' is of type 'bool' and has to be discarded
```
which is an ambiguous error. Should we just enforce that doc comments should begin immediately after the header?

**Comment by disruptek** [comment]
Unless I've missed something, doc comments are legal everywhere.  The body of the `proc` is a statement list, the `result` of which (pun intended) is the return value of the function; ie. the last statement or expression in the list.  This is how you are able to get away with not `discard`ing your result value in `foo`.  The subsequent syntax is thus incorrect for exactly the reason indicated.

**Comment by awr1** [comment]
> Unless I've missed something, doc comments are legal everywhere.

They are, but is there a good reason for this?

**Comment by disruptek** [comment]
Yes; the syntax is more consistent and thus easier for both the programmer and the compiler to reason about.  For example, I know why it shouldn't work and now you do too.
